### PR TITLE
Add a simple #[derive(JSManageable)] proc_macro

### PR DESCRIPTION
**Pull request by lqd**
@asajeffrey did you mean something like this ?

Since a macro 1.1 can't modify the item it's on, I went with the derive being on the NativeXXX instead of the XXX even though it would probably be better. 

Maybe a diesel-like macro like the following could be interesting:
```rust
linjs! {
    struct Console<'a, C> {
        window: Window<'a, C>,
    }
}
```

### CODE REVIEW

**Comment by asajeffrey** [comment]
Yes, that's a lot like what I was thinking of. Perhaps without the auto implementing the managed version, I quite like derive just to implement traits.

I'll have a proper look tomorrow, on my phone heading home right now!

**Comment by lqd** [comment]
Sure, let met know and I'll update the PR to remove the type alias!

File: Cargo.toml [hunk 1]
```diff
@@ -5,3 +5,6 @@ authors = ["ajeffrey@mozilla.com"]
 license = "MPL-2.0"
 
 [lib]
+
+[dependencies]
+linjs_derive = { path = "linjs-derive" }
```
Comment (asajeffrey) [inline Cargo.toml]: Should this dependency just be a test dependency?

Comment (asajeffrey) [reply Cargo.toml]: Nit: missing newline.

File: linjs-derive/lib.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+extern crate proc_macro;
+extern crate syn;
+
+#[macro_use]
+extern crate quote;
+
+use proc_macro::TokenStream;
+use syn::{MetaItem, Lit, Ident};
+
+#[proc_macro_derive(JSManageable, attributes(jsmanaged))]
```
Comment (asajeffrey) [inline linjs-derive/lib.rs]: I think remove the jsmanaged attribute.

File: linjs-derive/lib.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            }
+        }
+    }
+
+    let managed_name = managed_name.expect("the jsmanaged attribute is required, example: #[jsmanaged=\"ManagedAlias\"]");
+
+    quote! {
+        type #managed_name<'a, C> = JSManaged<'a, C, #name<'a, C>>;
+
+        unsafe impl<'a, 'b, C: JSCompartment> JSManageable<'b, C> for #name<'a, C> {
```
Comment (asajeffrey) [inline linjs-derive/lib.rs]: Deal with other type parameters? (I'm happy to assume exactly one lifetime param, and that the compartment is the first type param.)

Comment (asajeffrey) [reply linjs-derive/lib.rs]: This might be a separate PR though.

File: src/lib.rs [hunk 4]
```diff
@@ -170,6 +170,9 @@
 //!     Test.start();
 //! ```
 
+#[macro_use]
+extern crate linjs_derive;
```
Comment (asajeffrey) [inline src/lib.rs]: Just when `cfg(feature = "test")`?

**Review by asajeffrey — COMMENTED** [review]
Yay, looks good, some comments. I think we should just have the derive implement the JSManageable trait.

**Comment by lqd** [comment]
@asajeffrey alright I'm on it :)

Regarding the dev dependency and test cfg, this is to clean up the warning when compiling normally ? I think doing that would only allow linjs to use the macro, but not users of linjs, right ?

IIRC the setup right now is basically the same pattern as your deep-clone proc macro (which has the same unused warning). In any case it's no problem, I'll update the PR soon!

**Comment by asajeffrey** [comment]
Thanks! I'd like to make it possible to use linkjs without the macro, though this isn't really urgent.

**Comment by lqd** [comment]
@asajeffrey Updated with the requested fixes + handling of generic types

(btw did you see my other PR with doc fixes ?)

**Comment by lqd** [comment]
And we can also relax the "single-lifetime not named 'b" and "the JSCompartment is the first type param" constraints later if need be.

**Review by asajeffrey — APPROVED** [review]
LGTM!

**Pull request merged**